### PR TITLE
Introduce new `unmanaged-cluster` model

### LINKED ISSUE #2266: Offer minimal deployment model that supports development and experimentation of Tanzu
## Asks

1. Read this proposal
2. [Try out](https://github.com/vmware-tanzu/community-edition/blob/standalone-overhaul/cli/cmd/plugin/standalone-cluster/README.md) the proposed model
3. Vote 👍  or 👎  on this issue
4. If you have additional feedback, respond to this issue

We aim to `accept` or `reject` this proposal 60 days after opening (`12/17/2021`).

## Proposal

⚠️ The name of this feature is currently being determined, you may see it referred to as `local`, `standalone`, and `unmanaged-cluster` in various places. We have decided to move forward with the name `unmanaged-cluster`. Please note that references to `local` or `standalone` (not `standalone-cluster`) represent `unmanaged-cluster`

🚨  This proposal has been partially implemented to help further the conversation around whether we should accept it in this project. [Read here](https://github.com/vmware-tanzu/community-edition/blob/standalone-overhaul/cli/cmd/plugin/standalone-cluster/README.md) for details on how to try it and design details. 🚨

Standalone clusters (SAC) are our attempt to provide workload clusters without the need of a long-running management cluster. With this, we intended to:

* Minimize resource requirements for clusters
* Reduce on boarding time from download to cluster

To accomplish this, we re-purposed Cluster API and extended TKG-Lib (via tanzu-framework) to create the standalone cluster model. With this model in use for many months we've learned that:

1. SAC users often wanted a single-node or local cluster.
1. Users wanting more complex cluster-lifecycle management, need managed clusters.
1. Our usage of cluster-api and tkg-lib created a poor ux for `1` (above) and a stunted experience for `2` (above).
1. Maintaining the functionality support in our dependencies has had significant overhead as it's a use case tangential to our dependencies.

We believe that solving `1` is high-value for those using Tanzu.  We also believe that attempting to replicate cluster-lifecycle management on a single node comes at an inappropriate cost (via dependencies).

**We propose the deprecation of `standalone-cluster` in favor of introducing `local` (clusters).**

## High-level implementation details

In Tanzu, a Management Cluster does the processing of a `TanzuKubernetesRelease` (TKR). It uses the TKR to determine how to create a workload cluster.  

In the `local` model, we'll move the management cluster’s TKR processing client-side. After processing the TKR, we have all the information needed to create a Tanzu [workload] cluster that looks similar to that of one created by a management-cluster. See the following depiction of this relationship.

image

As seen above, after parsing the TKR (client-side) and understanding properties of the to-be-created cluster, we can call into a local provider to create a minimal cluster. By leveraging a provider abstraction (interface), we can insulate ourselves from the underlying details of how the infra/host/desktop-env are created. What matters is we receive a kubeconfig with admin access to the API server.

Our initial provider implementation will be `kind` because it's widely accepted in the Kubernetes community. The following gif demonstrates the bootstrap UX for a local Tanzu cluster.

_GIFs are broken up to save file-size_

### Cluster list / deletion:

`local` is good for:

* Workstation use cases of Tanzu
* Minimal (non-prod) resourced use cases of Tanzu
* CI/CD validations atop Tanzu
* Validation of newly created TKRs

`local` is not meant for:

* Cluster-lifecycle simulation
  * For this, use `management-cluster`.

Additionally, this approach would inherently solve many issues we face today:

* Resource utilization would be **drastically reduced**.
* WSL/Docker Desktop could be natively supported.
* Workstation/Host restarts would be supported.
* Added support for running multiple clusters.
* Testing of different component versions (e.g. kapp-controller) would be possible.
* We would delete our fork of `cluster-api` and `tanzu-framework`.

For in-depth implementation details, [please see our PR](https://github.com/vmware-tanzu/community-edition/pull/2257).

## Release Plan

* `0.10.0` will feature this new model alongside the existing `standalone-cluster` model.
* In `0.10.0`, the existing `standalone-cluster` model will print a deprecation notice to the user.
* In `0.11.0`, we'll remove the existing `standalone-cluster` model.

## FAQ

_This section will be updated as questions come in_

* **Q: What about non-local standalone-clusters (AWS, Azure, vSphere)**
  * A: Execution of this proposal **will** cause a gap for AWS, Azure, and vSphere as there will no longer be non-managed clusters available. However, `standalone-cluster`s are essentially very-limited management clusters with a few components ripped out. For users wanting to test and deploy a *single* cluster in one of those environments, we encourage simply creating a management-cluster and scheduling workloads to it. This **is not** our production-ready advise, but can get you the exact functionality (plus some) of the existing standalone-cluster model.

### PR DESCRIPTION
**Pull request by joshrosso**
## What this PR does / why we need it

See [this page to test this PR and understand details around its design](https://github.com/vmware-tanzu/community-edition/blob/standalone-overhaul/cli/cmd/plugin/standalone-cluster/README.md).

Please [add all feedback around this proposal to the proposal issue](https://github.com/vmware-tanzu/community-edition/issues/2266).

Please keep this PR exclusive to feedback about code.

Fixes: https://github.com/vmware-tanzu/community-edition/issues/2266

### CODE REVIEW

File: cli/cmd/plugin/unmanaged-cluster/main.go
Comment (joshrosso) [inline cli/cmd/plugin/unmanaged-cluster/main.go:19]: @stmcginnis @jpmcb I'd like to propose we make our aliases:

* `uc`
* `unmanaged`

Does that seem sensible to ya'll?

Comment (stmcginnis) [reply cli/cmd/plugin/unmanaged-cluster/main.go:19]: Those looks good, but I would also argue for `um`. That would be my default assumption based on the name.

**Comment by joshrosso** [comment]
@stmcginnis The tab writer we brought in from framework, it bolds the column headers of output, which I don't believe we should have for this plugin. While it is possible to work around this with OS detection, it still causes issue in many context such as using tools like `watch` where you get:

image

Are you or @jpmcb able to remedy this? I'm perfectly fine with copying the code over from framework.

**Comment by stmcginnis** [comment]
Yep, I'll address that today.

File: cli/cmd/plugin/unmanaged-cluster/main.go
Comment (jpmcb) [reply cli/cmd/plugin/unmanaged-cluster/main.go:19]: I agree with Sean:
- `um`
- `uc`
- `unmanaged`

Comment (joshrosso) [reply cli/cmd/plugin/unmanaged-cluster/main.go:19]: sgtm, can someone make this change in a PR?

**Comment by joshrosso** [comment]
We're getting close, so marking this as ready for review. A few known things that need to happen:

* [ ] @stmcginnis, @joshrosso, @jpmcb go through another full review
* [ ] @joshrosso communicates the proposal is accepted to the mailing list
* [ ] existing comments/requests for changes are addressed

Documentation is also a pre-req for release in v0.10.0, but I want to keep that to a separate issue/PR.

File: cli/cmd/plugin/unmanaged-cluster/main.go
Comment (stmcginnis) [reply cli/cmd/plugin/unmanaged-cluster/main.go:19]: On it.

**Comment by jpmcb** [comment]
@dvonthenen - looks like there's a merge conflict in the `hack/release/package-release.sh` script. My gut feeling is that chunk just needs to be removed as it was moved lower, but want to confirm with you

File: cli/cmd/plugin/standalone-cluster/create.go
Comment (joshrosso) [inline cli/cmd/plugin/standalone-cluster/create.go:52]: @jpmcb can you update this deprecation notice updated to mention the use of `unmanaged-cluster` since it will ship alongside `standalone-cluster`

File: cli/cmd/plugin/standalone-cluster/delete.go
Comment (joshrosso) [inline cli/cmd/plugin/standalone-cluster/delete.go:46]: @jpmcb can you update this deprecation notice updated to mention the use of `unmanaged-cluster` since it will ship alongside `standalone-cluster`

File: cli/cmd/plugin/standalone-cluster/main.go
Comment (joshrosso) [inline cli/cmd/plugin/standalone-cluster/main.go:17]: @jpmcb can you update this deprecation notice updated to mention the use of `unmanaged-cluster` since it will ship alongside `standalone-cluster`

File: cli/cmd/plugin/standalone-cluster/create.go
Comment (jpmcb) [reply cli/cmd/plugin/standalone-cluster/create.go:52]: @joshrosso - how's this wording look?

Screen Shot 2022-01-10 at 11 06 07 AM

Comment (jpmcb) [reply cli/cmd/plugin/standalone-cluster/create.go:52]: FYI - https://github.com/vmware-tanzu/community-edition/pull/2795

Comment (joshrosso) [reply cli/cmd/plugin/standalone-cluster/create.go]: love it. going to make one small recommendation that I missed (will do it in that PR.

File: cli/cmd/plugin/unmanaged-cluster/README.md
Comment (jpmcb) [inline cli/cmd/plugin/unmanaged-cluster/README.md:27]: These Google storage links appear broken and failing now in the URL linter. When I click the link, I see:

Screen Shot 2022-01-10 at 3 12 36 PM

**Review by jpmcb — CHANGES_REQUESTED** [review]

File: cli/cmd/plugin/unmanaged-cluster/README.md
Comment (jpmcb) [reply cli/cmd/plugin/unmanaged-cluster/README.md:27]: The binaries appear to still be at:

https://storage.googleapis.com/tce-tanzu-cli-plugins/build-tools/standalone/linux/tanzu-standalone

do we want to build some new ones and put them in the right bucket?

Comment (joshrosso) [reply cli/cmd/plugin/unmanaged-cluster/README.md:27]: > Or should we remove this chunk in the README? We probably don't need a pre-built binary of it if it's in the main release and on the main branch

My vote is to remove it in the README. Totally agreed that the pre-built binary isn't required and since the RFC window has close, folks can either:

1. get it in the v0.10.0 release
2. build from source

thoughts?

**Comment by joshrosso** [comment]
@stmcginnis @jpmcb this is officially open for review.

At a minimum, can you both build and validate:

* cluster creation
* cluster configuration
* cluster creation with configuration (some minimal settings tweaked)
* cluster deletion
* cluster listing

based on results of the above and/or any code review please:

* block and request changes
* approve for merge

Lets get 2 approvals before merge.

**Comment by joshrosso** [comment]
The following testing was performed after running `make install-all-tanzu-cli-plugins ENVS=linux-amd64` off 9cf7e3cac08b91997d1e031033547ec9492fad48. I am running on a system using cgroupsv2

### Created vanilla cluster :heavy_check_mark: 

```
$ tanzu unmanaged-cluster create doggo

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   Downloaded to: /home/josh/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/josh/.config/tanzu/tkg/unmanaged/doggo/config.yaml
   Bootstrap Logs: /home/josh/.config/tanzu/tkg/unmanaged/doggo/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster doggo
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/josh/.config/tanzu/tkg/unmanaged/doggo/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   antrea.tanzu.vmware.com:0.13.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to doggo

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete doggo
```

```
$ k get po -A
NAMESPACE            NAME                                          READY   STATUS    RESTARTS   AGE
kube-system          antrea-agent-jp5sk                            2/2     Running   0          75s
kube-system          antrea-controller-68b8dfb8d6-tdm5k            1/1     Running   0          74s
kube-system          coredns-558bd4d5db-tt7xg                      1/1     Running   0          2m16s
kube-system          coredns-558bd4d5db-v5jrv                      1/1     Running   0          2m16s
kube-system          etcd-doggo-control-plane                      1/1     Running   0          2m20s
kube-system          kube-apiserver-doggo-control-plane            1/1     Running   0          2m20s
kube-system          kube-controller-manager-doggo-control-plane   1/1     Running   0          2m20s
kube-system          kube-proxy-rcmv5                              1/1     Running   0          2m16s
kube-system          kube-scheduler-doggo-control-plane            1/1     Running   0          2m20s
local-path-storage   local-path-provisioner-547f784dff-cz8vw       1/1     Running   0          2m16s
tkg-system           kapp-controller-f7dbbcb4f-n2d9x               1/1     Running   0          2m16s

```

### Create configuration file :heavy_check_mark: 

```
$ tanzu uc configure newtest
Wrote configuration file to: newtest.yaml
```

```
$ cat newtest.yaml
ClusterName: newtest
KubeconfigPath: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: antrea
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
PortsToForward: []
SkipPreflight: false
```

### list clusters :heavy_check_mark: 

```
$ tanzu uc ls
  NAME   PROVIDER
  doggo  kind
```

### delete clusters with tty off :heavy_check_mark: 

```
$ tanzu uc rm doggo --tty-disable

 Deleting cluster: doggo

 Deleted cluster: doggo
```

### deployed cluster with calico set as cni in config

```
$ cat newtest.yaml
ClusterName: boooooop
KubeconfigPath: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: calico
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
PortsToForward: []
SkipPreflight: false
```

note `boooooop`

note `cluster name` is boooop and calico is selected

```
📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /home/josh/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/josh/.config/tanzu/tkg/unmanaged/boooooop/config.yaml
   Bootstrap Logs: /home/josh/.config/tanzu/tkg/unmanaged/boooooop/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster boooooop
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/josh/.config/tanzu/tkg/unmanaged/boooooop/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.tanzu.vmware.com:3.11.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to boooooop

```

```
$ k get po -A | grep -i calico
kube-system          calico-kube-controllers-cb5567d94-pnk84          0/1     Pending    0          2s
kube-system          calico-node-g5d4p                                0/1     Init:0/3   0          2s
```

**Comment by stmcginnis** [comment]
With `make install-all-tanzu-cli-plugins ENVS=darwin-amd64`

✅ cluster creation
```
tanzu um create choochoo

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   Downloading to: /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5.          Downloading to: /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5..         Downloaded to: /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /Users/smcginnis/.config/tanzu/tkg/unmanaged/choochoo/config.yaml
   Bootstrap Logs: /Users/smcginnis/.config/tanzu/tkg/unmanaged/choochoo/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster choochoo
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /Users/smcginnis/.config/tanzu/tkg/unmanaged/choochoo/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   antrea.tanzu.vmware.com:0.13.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to choochoo

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete choochoo
```

✅ cluster configuration
```
❯ tanzu um configure -c calico chugchug
Wrote configuration file to: chugchug.yaml

❯ bat chugchug.yaml
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: chugchug.yaml
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ ClusterName: chugchug
   2   │ KubeconfigPath: ""
   3   │ NodeImage: ""
   4   │ Provider: kind
   5   │ ProviderConfiguration: {}
   6   │ Cni: calico
   7   │ CniConfiguration: {}
   8   │ PodCidr: 10.244.0.0/16
   9   │ ServiceCidr: 10.96.0.0/16
  10   │ TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
  11   │ PortsToForward: []
  12   │ SkipPreflight: false
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

✅ cluster creation with configuration (some minimal settings tweaked)
```
❯ tanzu um create -f chugchug.yaml

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /Users/smcginnis/.config/tanzu/tkg/unmanaged/chugchug/config.yaml
   Bootstrap Logs: /Users/smcginnis/.config/tanzu/tkg/unmanaged/chugchug/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster chugchug
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /Users/smcginnis/.config/tanzu/tkg/unmanaged/chugchug/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.tanzu.vmware.com:3.11.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to chugchug

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete chugchug
```

✅ cluster listing
```
❯ tanzu um ls
  NAME      PROVIDER
  choochoo  kind
  chugchug  kind
❯ tanzu um ls -o yaml
- name: choochoo
  provider: kind
- name: chugchug
  provider: kind
❯ tanzu um ls -o json
[
  {
    "name": "choochoo",
    "provider": "kind"
  },
  {
    "name": "chugchug",
    "provider": "kind"
  }
]
❯ tanzu um ls -o json | jq '.[] | .name'
"choochoo"
"chugchug"
```

✅ cluster deletion
```
❯ tanzu um rm chugchug

🧪 Deleting cluster: chugchug

🧪 Deleted cluster: chugchug
❯ tanzu um rm choochoo

🧪 Deleting cluster: choochoo

🧪 Deleted cluster: choochoo
```

**Review by stmcginnis — DISMISSED** [review]

**Comment by jpmcb** [comment]
✅ cluster creation

```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 create meow

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/meow/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/meow/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster meow
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/meow/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   antrea.tanzu.vmware.com:0.13.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to meow

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete meow
```

cluster configuration
```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 configure -c calico woofwoof
Wrote configuration file to: woofwoof.yaml
❯ cat woofwoof.yaml
ClusterName: woofwoof
KubeconfigPath: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: calico
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
PortsToForward: []
SkipPreflight: false

```

✅ cluster creation with configuration (some minimal settings tweaked)

```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 create -f woofwoof.yaml woof

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/woof/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/woof/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster woof
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/woof/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.tanzu.vmware.com:3.11.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to woof

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete woof
```

✅ cluster listing
```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls
  NAME         PROVIDER
  hello-world  kind
  meow         kind
  meow01       kind
  woof         kind

❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls -o json
[
  {
    "name": "hello-world",
    "provider": "kind"
  },
  {
    "name": "meow",
    "provider": "kind"
  },
  {
    "name": "meow01",
    "provider": "kind"
  },
  {
    "name": "woof",
    "provider": "kind"
  }
]
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls -o yaml
- name: hello-world
  provider: kind
- name: meow
  provider: kind
- name: meow01
  provider: kind
- name: woof
  provider: kind
```

✅ cluster deletion
```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls -o json | jq '.[].name' | xargs -I _ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 delete _

🧪 Deleting cluster: hello-world

🧪 Deleted cluster: hello-world

🧪 Deleting cluster: meow

🧪 Deleted cluster: meow

🧪 Deleting cluster: meow01

🧪 Deleted cluster: meow01

🧪 Deleting cluster: woof

🧪 Deleted cluster: woof

```

**Review by jpmcb — DISMISSED** [review]
Let's do this 🚀

**Pull request merged**
